### PR TITLE
MODORDERS-299 Re-integrate with encumbrance API

### PR DESCRIPTION
**Pull request by pautau-s**
## Purpose
[MODORDERS-299](https://issues.folio.org/browse/MODORDERS-299) Re-integrate with encumbrance API

## Approach
if fundDistributions not empty
- for each `fundId` from `fundDistributions` fetching `curentFiscalYear` to built Transaction record
- creating order-transaction-summary wit numTransactions = fundDistributions.size()
- create Transaction records

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/rest/impl/FinanceHelper.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        .distinct()
+        .collect(toList()))
+      .map(this::getFundsByIds)
+      .toList()).thenApply(
+          lists -> lists.stream()
+            .flatMap(Collection::stream)
+            .collect(HashMap::new, accumulator(groupedByFund), Map::putAll));
+  }
+
+  @NotNull
```
Comment (khandramai) [inline src/main/java/org/folio/rest/impl/FinanceHelper.java]: Is this annotation necessary here?

Comment (pautau-s) [reply src/main/java/org/folio/rest/impl/FinanceHelper.java]: No. Maybe it was generated by IDE during refactoring

File: src/main/java/org/folio/rest/impl/FinanceHelper.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+      });
+  }
+
+  private boolean isFiscalYearNotFound(Throwable t) {
+    return t.getCause() instanceof HttpException && ((HttpException) t.getCause()).getCode() == 404;
+  }
+
+  private Transaction buildEncumbrance(FundDistribution distribution, CompositePoLine poLine, CompositePurchaseOrder compPo) {
+    MonetaryAmount estimatedPrice = calculateEstimatedPrice(poLine.getCost());
+    Transaction encumbrance = new Transaction();
```
Comment (KVupp) [inline src/main/java/org/folio/rest/impl/FinanceHelper.java]: we should be setting the other transaction fields like transactionType etc..

File: descriptors/ModuleDescriptor-template.json [hunk 3]
```diff
@@ -101,8 +103,10 @@
             "orders-storage.reporting-codes.item.put",
             "orders-storage.reporting-codes.item.delete",
             "configuration.entries.collection.get",
-            "finance-storage.encumbrances.collection.get",
-            "finance-storage.encumbrances.item.post",
+            "finance.funds.collection.get",
```
Comment (KVupp) [inline descriptors/ModuleDescriptor-template.json]: the POST is missing the funds permission and PUT is missing the transaction permission

File: src/main/java/org/folio/rest/impl/FinanceHelper.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+  FinanceHelper(HttpClientInterface httpClient, Map<String, String> okapiHeaders, Context ctx, String lang) {
+    super(httpClient, okapiHeaders, ctx, lang);
+  }
+
+  /**
+   * Creates Encumbrance records associated with given PO line and updates PO line with corresponding links.
+   *
+   * @return CompletableFuture with void on success.
+   */
+  CompletableFuture<Void> handleEncumbrances(CompositePurchaseOrder compPo) {
```
Comment (KVupp) [inline src/main/java/org/folio/rest/impl/FinanceHelper.java]: shouldn't we check if fund distributions are present before all the processing?

File: descriptors/ModuleDescriptor-template.json (hunk 3 above)
Comment (KVupp) [reply descriptors/ModuleDescriptor-template.json]: I don't think the transactions permission is needed. so just edit the POST permission

File: src/main/java/org/folio/rest/impl/FinanceHelper.java (hunk 4 above)
Comment (pautau-s) [reply src/main/java/org/folio/rest/impl/FinanceHelper.java]: This check was in different place but this method seems preferable for it.

File: descriptors/ModuleDescriptor-template.json [hunk 5]
```diff
@@ -40,8 +40,10 @@
             "orders-storage.reporting-codes.item.post",
             "orders-storage.reporting-codes.item.put",
             "configuration.entries.collection.get",
-            "finance-storage.encumbrances.collection.get",
-            "finance-storage.encumbrances.item.post",
+            "finance.encumbrances.item.post",
+            "finance.funds.item.get",
```
Comment (KVupp) [inline descriptors/ModuleDescriptor-template.json]: this should be funds collection permissions?

File: src/main/java/org/folio/rest/impl/FinanceHelper.java [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+    return t.getCause() instanceof HttpException && ((HttpException) t.getCause()).getCode() == 404;
+  }
+
+  private Transaction buildEncumbrance(FundDistribution distribution, CompositePoLine poLine, CompositePurchaseOrder compPo) {
+    MonetaryAmount estimatedPrice = calculateEstimatedPrice(poLine.getCost());
+    Transaction transaction = new Transaction();
+    transaction.setTransactionType(Transaction.TransactionType.ENCUMBRANCE);
+    transaction.setFromFundId(distribution.getFundId());
+    transaction.setSource(Transaction.Source.USER);
+    transaction.setCurrency(poLine.getCost().getCurrency());
```
Comment (KVupp) [inline src/main/java/org/folio/rest/impl/FinanceHelper.java]: The transaction currency must be the system currency , better to set it from the FY currency
> currency - currency code for this transaction - from the system currency.

Comment (pautau-s) [reply src/main/java/org/folio/rest/impl/FinanceHelper.java]: Should we than do conversion? Because for amount calculation  we use `poLine.cost,currency`.

**Pull request merged**